src/java/org/apache/cassandra/db/Keyspace.java

```diff
@@ -299,15 +301,34 @@ public boolean snapshotExists(String snapshotName)
     /**
      * Clear all the snapshots for a given keyspace.
      *
-     * @param snapshotName the user supplied snapshot name. It empty or null,
+     * @param snapshotName the user supplied snapshot name. If empty or null,
      *                     all the snapshots will be cleaned
+     * @param keyspace keyspace to remove snapshots for
      */
     public static void clearSnapshot(String snapshotName, String keyspace)
     {
+        Set<TableSnapshot> snapshotsToClear = new SnapshotLoader().loadSnapshots(keyspace)
+                                                                  .stream()
+                                                                  .filter(ts -> {
+                                                                      if (snapshotName == null || snapshotName.isEmpty())
+                                                                          return true;
+                                                                      return ts.getTag().equals(snapshotName);
+                                                                  })
+                                                                  .filter(ts -> {
+                                                                      if (snapshotName != null && !snapshotName.isEmpty() && ts.isEphemeral())
+                                                                          logger.info("Skipping deletion of ephemeral snapshot '{}' in keyspace {}. " +
+                                                                                      "Ephemeral snapshots are not removable by a user.",
+                                                                                      snapshotName, keyspace);
+
+                                                                      return !ts.isEphemeral();
+                                                                  })
+                                                                  .collect(Collectors.toSet());
+
         RateLimiter clearSnapshotRateLimiter = DatabaseDescriptor.getSnapshotRateLimiter();
 
-        List<File> tableDirectories = Directories.getKSChildDirectories(keyspace);
-        Directories.clearSnapshot(snapshotName, tableDirectories, clearSnapshotRateLimiter);
+        for (TableSnapshot snapshot : snapshotsToClear)
+            for (File snapshotDirectory : snapshot.getDirectories())
+                Directories.removeSnapshotDirectory(clearSnapshotRateLimiter, snapshotDirectory);
```

Comment:
It seems like this logic is duplicated on `SnapshotManager.clearSnapshot`. We also need to ensure expiring snapshots are properly removed from `SnapshotManager`. I think this logic is in `Keyspace` for historical reasons, now we can move it to `StorageService` so we can properly centralize snapshot clearing via `SnapshotManager. I created [this PR](https://github.com/instaclustr/cassandra/pull/47) moving `Keyspace.clearSnapshot` to `StorageService.clearKeyspaceSnapshot`. Let me know what do you think.